### DISCUSSION TITLE
How do I combine Dependency Injection and State in a Durable Entity?

### DISCUSSION THREAD

**Discussion by j2jensen** [Azure/azure-functions-durable-extension#2177] [Q&A]
[The docs](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-dotnet-entities#custom-initialization-on-first-access) encourage the use of this pattern to do custom initialization of a property prior to calling `DispatchAsync`.

```
[FunctionName(nameof(Counter))]
public static Task Run([EntityTrigger] IDurableEntityContext ctx)
{
    if (!ctx.HasState)
    {
        ctx.SetState(...);
    }
    return ctx.DispatchAsync<Counter>();
}
```

[This issue](https://github.com/Azure/azure-functions-durable-extension/issues/1563) indicates the problem with that approach: the object we pass in [must be](https://github.com/Azure/azure-functions-durable-extension/issues/1563#issuecomment-728781174) the entity instance itself. In experimenting it appears that calling `DispatchAsync` after calling `SetState` causes `DispatchAsync` to use the same instance of the entity that we passed in. That means we miss out on the Dependency Injection that `DispatchAsync` would normally perform.

[This comment](https://github.com/Azure/azure-functions-durable-extension/issues/1563#issuecomment-742151422) clarifies:

> Essentially, if you call any operations on `IDurableEntityContext` that edit state, you bypass the fancy reflection magic logic used by `DispatchAsync()`. The `DispatchAsync()` API was not designed with performing entity actions before dispatching in mind, so I would highly recommend against any business logic happening before you call `DispatchAsync()`. That static function should include exclusively boilerplate. In fact, it may be worth adding this case to our analyzer to highlight cases that could result in strange behavior.

That seems to directly contradict the pattern mentioned earlier. That [same comment](https://github.com/Azure/azure-functions-durable-extension/issues/1563#issuecomment-742151422) goes on to recommend passing the context itself into `DispatchAsync` and injecting it in the entity's constructor:

```
public FooEntity(... IDurableEntityContext ctx,...)
{
    if (!ctx.HasState)
    {
       ...
```

However, [this comment](https://github.com/Azure/azure-functions-durable-extension/issues/1162#issuecomment-573811082) points out that the static `Run` method is skipped entirely when someone calls `client.ReadEntityStateAsync`.

> All operations that are performed on the actual entity object get DI. So, you are guaranteed to get all the logging and all the configuration for all the operations performed on the "actual master copy" of the entity object.
>
> The case where you don't get DI are exactly the cases where the object is not the "actual master copy" of the entity and in that case you may need to be careful with logging or updating the object.
>
> For example, the state returned by client.ReadEntityStateAsync is a "readonly copy" of the entity state, not the actual entity state. ...

So calling `ReadEntityStateAsync` would cause the previous example to throw a null reference exception, because a `null` value would be passed into the `ctx` parameter. The same would be true if we just used `Entity.Current` instead of an injected context: that property is `null` when calling `ReadEntityStateAsync`.

So it seems like all the recommended patterns break in one of the following cases:
- Using information from the context to affect the State of the entity (e.g. one-time initialization)
- Using dependency injection to inject services into the Entity
- Getting a read-only copy of the object (e.g. `client.ReadEntityStateAsync`)

Are there recommended patterns that enable all of these cases to work together? Are they documented anywhere?

**Comment by davidmrdavid** [comment]
Hi @j2jensen,

Thanks for reaching out. After reviewing the links you shared, it does seem like we have some pending work to update our guidance here and frankly I'm also a bit confused on the best practices :-) . @sebastianburckhardt: any chance you chime in here with some advice?

I also see this old PR attempted to update our guidance but got closed prematurely: https://github.com/MicrosoftDocs/azure-docs-pr/pull/165981

@sebastianburckhardt: was this docs PR close to completion? Does it make sense to re-open it and wrap it up?
